src/basic_crawler.js

```diff
@@ -228,101 +229,102 @@ export default class BasicCrawler {
-                            return source.markRequestHandled(request);
-                        }
-
-                        if (!this.isStopped) request.pushErrorMessage(error);
-
-                        // Retry request.
-                        if (request.retryCount < this.maxRequestRetries) {
-                            if (!this.isStopped) {
-                                request.retryCount++;
-                                log.exception(error, 'BasicCrawler: handleRequestFunction failed, reclaiming failed request back to the list or queue', { // eslint-disable-line max-len
-                                    url: request.url,
-                                    retryCount: request.retryCount,
-                                });
-                            }
-                            willBeRetried = true;
-                            return source.reclaimRequest(request);
-                        }
-
-                        if (!this.isStopped) {
-                            log.exception(error, 'BasicCrawler: handleRequestFunction failed, marking failed request as handled', {
-                                url: request.url,
-                                retryCount: request.retryCount,
-                            });
-                        }
-
-                        // Mark as failed.
-                        return source
-                            .markRequestHandled(request)
-                            .then(() => this.handleFailedRequestFunction({ request, error }));
-                    })
-                    .finally(() => {
-                        if (!willBeRetried) this.handledRequestsCount++;
-                    });
-            });
+        const request = await this._fetchNextRequest();
+        if (!request) return;
+
+        try {
+            // rejectOnStopPromise rejects when .stop() is called or AutoscaledPool throws.
+            // All running tasks are therefore terminated with an error to be reclaimed and retried.
+            await Promise.race([this.handleRequestFunction({ request }), this.rejectOnStopPromise]);
+            source.markRequestHandled(request);
+            this.handledRequestsCount++;
+        } catch (err) {
+            await this._requestFunctionErrorHandler(err, request, source);
+        }
     }
 
     /**
      * Returns true if some RequestList and RequestQueue have request ready for processing.
      *
      * @ignore
      */
-    _isTaskReadyFunction() {
-        return Promise
-            .resolve()
-            .then(() => {
-                if (!this.requestList) return true;
-
-                return this.requestList.isEmpty();
-            })
-            .then((isRequestListEmpty) => {
-                if (!isRequestListEmpty || !this.requestQueue) return isRequestListEmpty;
-
-                return this.requestQueue.isEmpty();
-            })
-            .then(areBothEmpty => !areBothEmpty);
+    async _isTaskReadyFunction() {
```

Comment:
It's much more readable now. But the potential overhead is that when you are using both request-list and request-queue then everytime you call this it calls `this.requestQueue.isEmpty()` which makes a request to request-queue API and therefore to dynamo DB. And this request is not needed because we know if request-list is nonempty immediately.

Reply:
Good point, I was wondering whether to do it separately or not. This did not occur to me.